### ISSUE TITLE
ioBroker.tado produces exceptionally high CPU load

### ISSUE THREAD

**Issue by paulheimann** [DrozmotiX/ioBroker.tado#192]
After installing ioBroker.tado I noticed the fan of my [Pi PoE HAT](https://www.raspberrypi.org/products/poe-hat/) starting to periodically cool the CPU (which it never did before). Based on the output of `htop` I noticed that ioBroker.tado is producing 40 - 50% CPU load with every refresh interval. 
Based on the silly log output it seems that the adapter re-creates all objects with every refresh. My tado setup is quite complex with 14 rooms and ~20 devices. 
This also increases the load of ioBroker.javascript at the same time as it seems its consuming the flood of events from ioBroker.tado. As soon as I disable my instance of ioBroker.tado the periodic CPU load increases stop (also for the ioBroker.javascript adapter) and the CPU cooling ends.

No other of my (50+) adapters is producing these high levels of load. 

**Expected behavior**  
The adapter should create less load on the CPU.

**Versions:**  
 - Adapter version: 0.2.6
 - JS-Controller version: 3.2.16
 - Node version: 14.16.0
 - Operating system: Raspbian GNU/Linux 10 (buster)

**Comment by HGlab01** [comment]
Hi! Which interval is configured?
I / we will have a look at the state update.

**Comment by DutchmanNL** [comment]
Hmm i See here we have a nasty extendObject at every refresh, that’s inefficient.

https://github.com/DrozmotiX/ioBroker.tado/blob/main/main.js#L1648

sorry old piece of code, already fixed in my other adapters will take care of this one during weekend

**Comment by HGlab01** [comment]
we could reuse the implemented function in json-explorer ;-)

**Comment by DutchmanNL** [comment]
> @DutchmanNL  we could reuse the implemented function in json-explorer ;-)

That's the one I mean 😉

**Comment by HGlab01** [comment]
> > @DutchmanNL  we could reuse the implemented function in json-explorer ;-)
> 
> That's the fix I mean 😉

I did a quick hack and it seems to work very fine
https://github.com/DrozmotiX/ioBroker.tado/tree/192
It just produces a lot of warnings, because stateAttr.js does not cover all states. But this is just something that has to be done ;-)

**Comment by HGlab01** [comment]
@paulheimann 
If you would like to do an experimental test you can install
DrozmotiX/ioBroker.tado#082a3ab2b8119a9eb2c1907ed6c40d0ee5f18c98
like
![image](https://user-images.githubusercontent.com/54232526/112749022-e9dc8180-8fbf-11eb-9488-a95db00cc16c.png)

You will get a lot of warnings like 
`Sentry disabled, error catched : State attribute definition missing for 'Stage_04_Weather'`
but this does not impact the functionality.

Would be interesting if CPU usage decreases.

**Comment by paulheimann** [comment]
@HGlab01 @DutchmanNL, thanks a lot for the rapid response! I'm impressed.

> Hi! Which interval is configured?

30 seconds, I believe that's the default which I never changed.

> If you would like to do an experimental test you can install
> DrozmotiX/ioBroker.tado#8cba3f19afb5654f95501786de2eda4bc4307766

I tried. What I observed: Load on ioBroker.tado goes down to 20 - 25% CPU with each update interval. The simultaneous load spikes of ioBroker.javascript seem to have gone down more significantly (~5 - 10%).

After having run ioBroker for ~1 hr, ioBroker.tado is still the sub process with by far the highest CPU time. So, while this change definitely improved the situation it still seems that ioBroker.tado is more resource hungry than it should be.

**Comment by DutchmanNL** [comment]
> @HGlab01 @DutchmanNL, thanks a lot for the rapid response! I'm impressed.
> 
> > Hi! Which interval is configured?
> 
> 30 seconds, I believe that's the default which I never changed.
> 
> > If you would like to do an experimental test you can install
> > DrozmotiX/ioBroker.tado#8cba3f19afb5654f95501786de2eda4bc4307766
> 
> I tried. What I observed: Load on ioBroker.tado goes down to 20 - 25% CPU with each update interval. The simultaneous load spikes of ioBroker.javascript seem to have gone down more significantly (~5 - 10%).
> 
> After having run ioBroker for ~1 hr, ioBroker.tado is still the sub process with by far the highest CPU time. So, while this change definitely improved the situation it still seems that ioBroker.tado is more resource hungry than it probably should be.

i will take a further look at that when I change the oAuth handling. Basically we only get a JSON and write it to states should not be a high load

**Comment by paulheimann** [comment]
Just a quick update: Kept iobroker running for more than than 24hrs with the experimental version from @HGlab01. In terms of CPU time ioBroker.tado is still by far the most hungry adapter (roughly 10x compared to the second most resource hungry one).

**Comment by HGlab01** [comment]
My Tado-Install has 216 states. Therefore it is not surprising when 200+ Events are fired every polling interval.
![image](https://user-images.githubusercontent.com/54232526/113018153-83529180-9180-11eb-8589-ccaaf0bb6d3f.png)

On my system Tado is comparable to the admin adapter (comparing CPU time).
![image](https://user-images.githubusercontent.com/54232526/113018402-cca2e100-9180-11eb-821c-e900a466e6a4.png)
Changing the interval from 30 sec to 120 sec would decrease the CPU load to a quarter.

**Comment by paulheimann** [comment]
> My Tado-Install has 216 states. Therefore it is not surprising when 200+ Events are fired every polling interval.

I'm absolutely no expert so please forgive if this is an irrelevant thought: Shouldn't events only be fired for things that actually changed compared to the last poll?

**Comment by DutchmanNL** [comment]
> > My Tado-Install has 216 states. Therefore it is not surprising when 200+ Events are fired every polling interval.
> 
> I'm absolutely no expert so please forgive if this is an irrelevant thought: Shouldn't events only be fired for things that actually changed compared to the last poll?

Hmm yes and now, there are 2 ways to do it : 

1) (as now) let the controller handle the state behavior
2) (possible to do) keep the values in memory and only write if the value changed

we can make a test what's more efficient, to be honest didn't compare it yet between the ioBroker controller logic or doing ut by ourself with memory logic

**Comment by HGlab01** [comment]
In my point of view this should not be done on adapter level but on framework (= iobroker js-controller) level. And I would say the current iobroker js-controller implementation is pretty fine.
I appreciate inputs and ways how to get it more efficient and we should proceed to think how to elaborate.

**Comment by DutchmanNL** [comment]
@HGlab01 i Will check with the Core team about backend and planner Roadmap on These items as the value are Updates (we cam See That) question is iFrame it only Updates the State Time stamp are anwatschle the valies amd ACK too

of the Last is the case, we have 2 unneeded write actions to the state DB

@paulheimann what kind of hardware is your ioBroker running on?

**Comment by paulheimann** [comment]
> @paulheimann what kind of hardware is your ioBroker running on?

Running on Raspberry Pi 4 Model B 4GB RAM

**Comment by DutchmanNL** [comment]
> > @paulheimann what kind of hardware is your ioBroker running on?
> 
> Running on Raspberry Pi 4 Model B 4GB RAM

what kind of storage ? On an raspberry pi with an SD you should really think about using Redis instead of writing on the card

**Comment by paulheimann** [comment]
> what kind of storage ? On an raspberry pi with an SD you should really think about using Redis instead of writing on the card

Yes, you are right. Unfortunately didn't find time yet to switch to redis yet.

**Comment by paulheimann** [comment]
Switched to redis in the meantime. iobroker.tado (running on the version  DrozmotiX/ioBroker.tado#8cba3f19afb5654f95501786de2eda4bc4307766 referenced by @HGlab01 above) is still the process with the highest CPU consumption on my entire system (monitoring it now for a couple of days). The CPU spikes with every update interval now go hand in hand with spikes of redis.

**Comment by HGlab01** [comment]
I compared the CPU-time to other adapters:
iobroker.javascript: 5-times
iobroker.shelly (also state intensiv): 2-times
iobroker.scene: ~equal

My assessment: CPU load is ok in comparison to other adapters For sure, if adapters do not handle many states the CPU consumption is less.

I will combine this improvement and issue 214 and launch a version for latest in the next few days

**Comment by HGlab01** [comment]
If you like you can check this version. This includes all possible improvements.
DrozmotiX/ioBroker.tado#d5e9999b562c02c63c16e24c5eeb22135a787992

**Comment by HGlab01** [comment]
Release Candidate:
DrozmotiX/ioBroker.tado#v0.2.7-2

**Comment by HGlab01** [comment]
improved in v0.2.7 (https://github.com/DrozmotiX/ioBroker.tado/commit/474dfc54caefa13c7c5fc68256290e5a68cb8c64)
